atools/_memoize_decorator.py

```diff
@@ -190,7 +193,29 @@ def finalize_memo(self, memo: _Memo, key: Union[int, str]) -> Any:
                 )
             return memo.memo_return_state.value
 
-    def get_hashed_key(self, key: Tuple[Hashable]) -> Union[int, str]:
+    def get_hashed_key(self, key: Union[Hashable, Tuple[Hashable]]) -> Union[int, str]:
+        # Make a best-effort attempt to warn developers if they use a keygen
+        # function that relies on the default hash implementation for object.
+        if not self._has_checked_keygen:
+            if isinstance(key, tuple):
+                for component in key:
+                    # The default  __hash__ implementation can re-use hash values
+                    # if the original object is deleted.
+                    if hasattr(type(component), "__hash__") and type(component).__hash__ == object.__hash__:
```

Comment:
I couldn't think of a way to detect if an object incorrectly has a non-unique hash, but we can detect if the class is using the known-bad case of the default `object.__hash__`.